{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Finder Write-Up & Reflection\n",
    "### Tolulope Oshinowo\n",
    "### Software Design Mini Project 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this project a gene finder was developed in Python with the intention to detect amino acids from entered DNA. It was a multi-faceted effort spanning two weeks, but by the end meaningful information was successfully derived. In this exercise the gene finder printed several possible candidate proteins from a given strand of Salmonella DNA, but of them one that stuck out was the \"SPI-1 type III secretion system protein (SpaO)\". According to the NCBI Protein Blast this protein was recognized as Salmonella Enterica, and based off of the BLAT Genome Finder it most likely came from a Bonobo or Gibbon. These genes were for the most part excretionary, and knowing that type III secretion systems are typically invasive it is safe to say that this protein is pathogenic. Below is the protein I decided to use from my Gene Fider output:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MGIIKIIIADNFDQSMADYSLAVFGIGLKYLIPFMLLCLVCSALPALLQAGFVLATEALKPNLSALNPVEGAKKLFSMRTVKDTVKTLLYLSSFVVAAIICWKKYKVEIFSQLNGNIVGIAVIWRELLLALVLTCLACALIVLLLDAIAEYFLTMKDMKMDKEEVKREMKEQEGNPEVKSKRREVHMEILSEQVKSDIENSRLIVANPTHITIGIYFKPELMPIPMISVYETNQRALAVRAYAEKVGVPVIVDIKLARSLFKTHRRYDLVSLEEIDEVLRLLVWLEEVENAGKDVIQPQENEVRH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aside from the quantitative aspects of this project, it is also important to look at the qualitative and ethical characteristics of the Gene Finder python script. In terms of the beneficial potential that this technology has for the general population, healthcare is one that quickly comes to mind. Thanks to gene finding technology diseases can be easily checked for and combated before more obvious symptoms begin to show fieldsâ€”effectively saving countless lives."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking forward there is still plenty of room for improvement in regards to my Gene Finder. For starters, the code is limited by the fact that it does not account for incorrect entries. This would mean that false positives and negatives could arise from mis-entering codons. Also in this program each method is derivative of the previous one with no way of cross-checking accuracy; meaning that crucial information could get lost in the mix. In the future I would adapt this program to work for more test cases and increase the safety factor across the board. The health industry plays no games, and if I want to one day be a successful developer in that field minimal margins of error are mandatory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### THE FULL CODE:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "GENE FINDER MINI PROJECT ONE\n",
    "\n",
    "@author: Tolulope Oshinowo\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "import random\n",
    "from amino_acids import aa, codons, aa_table   # you may find these useful\n",
    "from load import load_seq\n",
    "\n",
    "\n",
    "def shuffle_string(s):\n",
    "    return ''.join(random.sample(s, len(s)))\n",
    "\n",
    "\n",
    "def get_complement(nucleotide):\n",
    "    #Making sure that complements work properly.\n",
    "    \"\"\"\n",
    "    >>> get_complement(\"A\")\n",
    "    'T'\n",
    "    >>> get_complement(\"C\")\n",
    "    'G'\n",
    "    \"\"\"\n",
    "    if(nucleotide == 'A'):\n",
    "        return 'T'\n",
    "    elif(nucleotide == 'T'):\n",
    "        return 'A'\n",
    "    elif(nucleotide == 'C'):\n",
    "        return 'G'\n",
    "    elif(nucleotide == 'G'):\n",
    "        return 'C'\n",
    "    else:\n",
    "        return 'no nucleotide given'\n",
    "\n",
    "def get_reverse_complement(dna):\n",
    "    #Making sure that reverse strand operates properly.\n",
    "    \"\"\"\n",
    "    >>> get_reverse_complement(\"ATGCCCGCTTT\")\n",
    "    'AAAGCGGGCAT'\n",
    "    >>> get_reverse_complement(\"CCGCGTTCA\")\n",
    "    'TGAACGCGG'\n",
    "    \"\"\"\n",
    "    length = len(dna)\n",
    "    output = ''\n",
    "    while length > 0:\n",
    "        output = output + get_complement(dna[length-1:length])\n",
    "        length = length - 1\n",
    "    return output\n",
    "\n",
    "def rest_of_ORF(dna):\n",
    "    #Making sure start codon can be identified and the stop can get cut off in sets of 3.\n",
    "    \"\"\"\n",
    "    >>> rest_of_ORF(\"ATGTGAA\")\n",
    "    'ATG'\n",
    "    >>> rest_of_ORF(\"ATGAGATAGG\")\n",
    "    'ATGAGA'\n",
    "    \"\"\"\n",
    "    if(dna[0:3] == 'ATG'):\n",
    "        pos = 0\n",
    "        length = len(dna)\n",
    "        output = ''\n",
    "        while length > 0:\n",
    "            if(dna[pos:pos+3] == 'TAA'):\n",
    "                return output\n",
    "            if(dna[pos:pos+3] == 'TAG'):\n",
    "                return output\n",
    "            if(dna[pos:pos+3] == 'TGA'):\n",
    "                return output\n",
    "            output = output + dna[pos:pos+3]\n",
    "            pos = pos + 3\n",
    "            length = length - 3\n",
    "        return output\n",
    "    else:\n",
    "        return 'no start codon found'\n",
    "\n",
    "def find_all_ORFs_oneframe(dna):\n",
    "    #Making sure that the previous function can generate a list of codons can be returned.\n",
    "    \"\"\"\n",
    "    >>> find_all_ORFs_oneframe(\"ATGCATGAATGTAGATAGATGTGCCC\")\n",
    "    ['ATGCATGAATGTAGA', 'ATGTGCCC']\n",
    "    \"\"\"\n",
    "    pos = 0\n",
    "    full = len(dna)\n",
    "    length = len(dna)\n",
    "    string = ''\n",
    "    output = []\n",
    "    while length > 0:\n",
    "        string = rest_of_ORF(dna[pos:full])\n",
    "        if(dna[pos:pos+3] == 'ATG'):\n",
    "            output.append(string)\n",
    "        pos = pos + 3\n",
    "        string = ''\n",
    "        length = length - 3\n",
    "    return output\n",
    "\n",
    "def find_all_ORFs(dna):\n",
    "    #Making sure that the previous function can be used to parse through the stand for codons.\n",
    "    \"\"\"\n",
    "    >>> find_all_ORFs(\"ATGCATGAATGTAG\")\n",
    "    ['ATGCATGAATGTAG', 'ATGAATGTAG', 'ATG']\n",
    "    \"\"\"\n",
    "    pos = 0\n",
    "    full = len(dna)\n",
    "    length = len(dna)\n",
    "    string = ''\n",
    "    output = []\n",
    "    while length > 0:\n",
    "        output += find_all_ORFs_oneframe(dna)\n",
    "        dna = dna[full-(length-1):full]\n",
    "        length = length - 3\n",
    "    return output\n",
    "\n",
    "def find_all_ORFs_both_strands(dna):\n",
    "    #Making sure virtually all the prevous functions can come together to return ORFs of both DNA strands.\n",
    "    \"\"\"\n",
    "    >>> find_all_ORFs_both_strands(\"ATGCGAATGTAGCATCAAA\")\n",
    "    ['ATGCGAATG', 'ATGCTACATTCGCAT']\n",
    "    \"\"\"\n",
    "    reverse = get_reverse_complement(dna)\n",
    "    output = []\n",
    "    output += find_all_ORFs(dna)\n",
    "    output += find_all_ORFs(reverse)\n",
    "    i = len(output)-1\n",
    "    while i > -1:\n",
    "        if output[i] == 'ATG':\n",
    "            output.remove(output[i])\n",
    "        i-=1\n",
    "    return output\n",
    "\n",
    "def longest_ORF(dna):\n",
    "    #Making sure the longest ORF can be selected from a giveln list of ORFs.\n",
    "    \"\"\"\n",
    "    >>> longest_ORF(\"ATGCGAATGTAGCATCAAA\")\n",
    "    'ATGCTACATTCGCAT'\n",
    "    \"\"\"\n",
    "    strands = find_all_ORFs_both_strands(dna)\n",
    "    length = len(strands)\n",
    "    x = 0\n",
    "    i = 0\n",
    "    n = 0\n",
    "    strand = ''\n",
    "    for i in strands:\n",
    "        y = len(strands[n])\n",
    "        if y>x:\n",
    "            strand = strands[n]\n",
    "        n += 1\n",
    "    return strand\n",
    "\n",
    "def longest_ORF_noncoding(dna, num_trials):\n",
    "    #This method can not really be doctested due to randomization.\n",
    "    import random\n",
    "    max = 0\n",
    "    for i in range(num_trials):\n",
    "        newdna = ''.join(random.sample(dna, len(dna)))\n",
    "        if len(longest_ORF(newdna)) > max:\n",
    "            max = len(longest_ORF(newdna))\n",
    "    return max\n",
    "\n",
    "def coding_strand_to_AA(dna):\n",
    "    #Makes sure that amino acids are being properly transcribed.\n",
    "    \"\"\"\n",
    "    >>> coding_strand_to_AA(\"ATGCGA\")\n",
    "    'MR'\n",
    "    >>> coding_strand_to_AA(\"ATGCCCGCTTT\")\n",
    "    'MPA'\n",
    "    \"\"\"\n",
    "    length = len(dna)\n",
    "    pos = 0\n",
    "    aa = ''\n",
    "    while(length > 2):\n",
    "        aa += aa_table[dna[pos:pos+3]]\n",
    "        pos += 3\n",
    "        length -= 3\n",
    "    return aa\n",
    "\n",
    "def gene_finder(dna):\n",
    "    #Makes sure it all comes together\n",
    "    from load import load_seq\n",
    "    dna = load_seq(\"/home/tolu/GeneFinder/data/X73525.fa\")\n",
    "    threshold = longest_ORF_noncoding(dna, 1500)\n",
    "    output = []\n",
    "    protein = ''\n",
    "    orfs = find_all_ORFs_both_strands(dna)\n",
    "    i = 0\n",
    "    for i in range(len(orfs)):\n",
    "        if(len(orfs[i]) >= threshold):\n",
    "            protein = coding_strand_to_AA(orfs[i])\n",
    "            output.append(protein)\n",
    "    return output\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    import doctest\n",
    "    print(gene_finder(''))\n",
    "    '''\n",
    "    doctest.run_docstring_examples(get_complement, globals(), verbose=True)\n",
    "    doctest.run_docstring_examples(get_reverse_complement, globals(), verbose=True)\n",
    "    doctest.run_docstring_examples(rest_of_ORF, globals(), verbose=True)\n",
    "    doctest.run_docstring_examples(find_all_ORFs_oneframe, globals(), verbose=True)\n",
    "    doctest.run_docstring_examples(find_all_ORFs, globals(), verbose=True)\n",
    "    doctest.run_docstring_examples(find_all_ORFs_both_strands, globals(), verbose=True)\n",
    "    doctest.run_docstring_examples(longest_ORF, globals(), verbose=True)\n",
    "    #doctest.run_docstring_examples(longest_ORF_noncoding, globals(), verbose=True) randomization removes necessity of this test\n",
    "    doctest.run_docstring_examples(coding_strand_to_AA, globals(), verbose=True)\n",
    "    '''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
